infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,613 @@
+            log.warn("Attempted to complete request {}, but request not outstanding!", requestID);
+        }
+    }
+
+    @Override
+    public void completeExceptionally(LogReplicationSession session, long requestID, Throwable cause) {
+        CompletableFuture cf;
+        if ((cf = sessionToOutstandingRequests.get(session).remove(requestID)) != null) {
+            cf.completeExceptionally(cause);
+            log.debug("completeExceptionally: Remove request {} to {} due to {}.", requestID,
+                    sessionToRemoteClusterDescriptor.get(session).getClusterId(), cause.getClass().getSimpleName(), cause);
+        } else {
+            log.warn("Attempted to exceptionally complete request {}, but request not outstanding!",
+                    requestID);
+        }
+    }
+
+    @Override
+    public void stop(Set<LogReplicationSession> sessions) {
+        sessions.forEach(session -> removeSessionInfo(session));
+
+        // When all the connection starter sessions are stopped, shut down the transport layer
+        if(sessionToOutstandingRequests.isEmpty() && this.clientChannelAdapter != null) {
+            // stop the client Adapter. The Server Adapter is closed in the interClusterServerNode.
+            this.clientChannelAdapter.stop();
+        }
+    }
+
+    @Override
+    public void setTimeoutResponse(long timeoutResponse) {
+        this.timeoutResponse = timeoutResponse;
+    }
+
+    /**
+     * Interfaces between the LogReplicationServer (msg handler), and the transport layer
+     * @param response Log replication response message
+     */
+    @Override
+    public void sendResponse(CorfuMessage.ResponseMsg response) {
+        log.trace("Ready to send response {}", response.getPayload().getPayloadCase());
+        LogReplicationSession session = response.getHeader().getSession();
+        if (isConnectionInitiator(session)) {
+            if(sessionManager.getIncomingSessions().contains(session)) {
+                if (sessionToSinkVerifyLeadership.get(session).getRemoteLeaderNodeId().isPresent()) {
+                    clientChannelAdapter.send(sessionToSinkVerifyLeadership.get(session).getRemoteLeaderNodeId().get(), response);
+                } else {
+                    sessionToSinkVerifyLeadership.get(session).input(
+                            new LogReplicationSinkEvent(LogReplicationSinkEvent.LogReplicationSinkEventType.REMOTE_LEADER_LOSS));
+                }
+            }
+        } else {
+            try {
+                this.serverChannelAdapter.send(response);
+                log.trace("Sent response: {}", response);
+            } catch (IllegalArgumentException e) {
+                log.warn("Illegal response type. Ignoring message.", e);
+            }
+        }
+    }
+
+    /**
+     * Receive Corfu Message from the Channel Adapter for further processing
+     *
+     * @param msg received corfu message
+     */
+    @Override
+    public void receive(CorfuMessage.ResponseMsg msg) {
```

Comment:
Based on the nature of the msgs, all the msgs from the source will be requests, and all the msgs from sink are responses.

But we now have a constraint when sink is the connection initiator..since sink has to set a long living stream first, and source can then drive the replication on that stream.
To create a stream, we trigger an rpc...and rpc has the type of incoming and outgoing msg.
in this long living rpc, incoming msg will be `RequestMsgs` and outgoing msgs will be `responseMsgs`. 
To comply with this, the first msg to source in the reverseReplicate rpc is the subscribeMsg, and I have it as of type ResponseMsg.

If we don't want the lr_subscribe msg to be of "responseType" then we'll need to have another wrapper around msgs...which I think would be an overkill.
subscribe_msg on the source is like a signal for source to start its FSM